### PR TITLE
Coercion hook

### PR DESCRIPTION
**Pull request by proux01**
Draft pull request to track the coercion hook feature developped during CUDW 2023.

N.B.: currently the branch is based on coq-master but includes a merge with master so the diff is pretty unreadable, only the last two commits are actually part of the PR.

Requires Coq >= 8.18 and tested with Elpi master.

C.f. math-comp/math-comp#1051 for examples of use

### CODE REVIEW

**Comment by gares** [comment]
great, I'll look into is ASAP

**Comment by proux01** [comment]
No hurry

File: Changelog.md [hunk 1]
```diff
@@ -1,10 +1,38 @@
 # Changelog
 
-## UNRELEASED
+## Unreleased
```
Comment (gares) [inline Changelog.md:3]:
```suggestion
## Unreleased

Requires Elpi 1.16.5 and Coq 8.18.
```

File: apps/coercion/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+Elpi Typecheck Coercion.
+
+Check true : Prop.
+Check false : Prop.
+```
+
+## Requirements
+
+Appropriate version of Coq with support for coercion hooks
+https://github.com/coq/coq/pull/17794
```
Comment (gares) [inline apps/coercion/README.md:25]:
```suggestion
## Status

This app is experimental.
```

**Comment by gares** [comment]
Overall it is fine. I think that for uniformity and to avoid conflicts:
- `Coercion` -> `coercion` (the program name).
- `Elpi AddCoercionHook Coercion` -> `Elpi Coercion coercion` 
- The `coercion` predicate should be in a Db (eg coercion.db), which should be accumulated by the main program, possibly as the last operation. This makes it possible to share it with other apps. You should then accumulate to the Db, eg `Elpi Accumulate coercion.db lp:{{ ... }}`.

I'm wondering if the app should:
- allow to "unset" the hook (imagine the program contains a bad bug, and you want to disable/re-enable it)
- document a `Elpi Print coercion "coercion" "coercion".` as a way to see the current program, or have an ad-hoc `Elpi Coercion Print` command (doing that behind the scenes).
  
I'm also wondering if the .ml code should all be moved the the apps/coercion directory. I've never tried it, but (assuming all APIs from elpi_vernacular/programs are exported) makes a bit more sense (In some sense apps should not require vernaculars from the core). I can try this myself once this is merged.

**Comment by gares** [comment]
Bottom line, coq-master becomes master when the next rc is released. The requirements on Coq won't allow you to use this commit on a Coq which is not 8.18

**Comment by proux01** [comment]
Thanks for the review and sorry for the late answer.
* I agree and followed your renaming advices
* put coercions in a DB
* added command `Elpi Disable Coercion`
* documented how to print coercions (although filtering doesn't seem to work on coq-master so output is somewhat clutered until one uses the textbox in the webpage)

> I'm also wondering if the .ml code should all be moved [...]  I can try this myself once this is merged.

I agree, thanks

**Comment by gares** [comment]
I'm sorry I did the release including an old version of this PR.
Could you please tease te newcomits on master?

**Comment by gares** [comment]
I did do the requested changes I believe, but feel free to improve further

**Pull request closed without merging**